### PR TITLE
More robust staggered Jacobian and RGN_NOCORNERS

### PR DESCRIPTION
**Pull request by johnomotani**
Fabio found that when calculating the Jacobian `J` at `CELL_YLOW` it is more robust to interpolate `J` from the cell-centre `J`, rather than calculating `J` from the interpolated metric coefficients; calculating from interpolated metric coefficients resulted in a negative `J` at ylow near the X-point, while interpolating `J` itself gave a result that was always positive. This PR also treats `Bxy` similarly for consistency, as `Bxy` is calculated from `J`.

Also add a `RGN_NOCORNERS` to the default regions created in `Mesh::createDefaultRegions()`, as well as `RGN_XGUARDS`, `RGN_YGUARDS` and `RGN_ZGUARDS` which are used to build `RGN_NOCORNERS`. `RGN_NOCORNERS` includes all the x- and y-guard cells (internal and boundary) except the corner guard-cells. This is useful for checking in the `Coordinates` constructors as `interpolateAndExtrapolate` sets the corner cells to `BoutNaN`, but it's still useful to check the guard cells.
Re-name the region created in `invertable_operator.hxx` from `RGN_NOCORNERS` to `RGN_WITHBNDRIES` as this region excludes internal guard cells, and only includes the boundary guard-cells.

Added some utility functions `checkFinite(f)` and `checkPositive(f)` that do similar things to `finite(f)` and `min(f) > 0.` but throw exceptions instead of returning a bool. The exception messages include the indices of the point where the check failed - added `toString` functions for `Ind3D`, `Ind2D` and `IndPerp` to support this.

The changes to the Jacobian calculation mean that the `Coordinates::jacobian()` method uses the `GridDataSource` from `Mesh` (to check whether it's necessary to extrapolate into boundary guard cells). This broke the tests as `source` was `nullptr` for the mesh in `FakeMeshFixture` - add a `FakeGridDataSource` to `test_extras.hxx` to get around this. `FakeGridDataSource` does not actually supply data, the `get(...)` methods all just return `false`, but gets around the segfault in `jacobian()`. It might be nice at some point to use `FakeGridDataSource` to actually supply `g11`, etc. so that we could unit-test the `Coordinates` constructors, but that's beyond the scope of this PR...

### CODE REVIEW

**Comment by johnomotani** [comment]
Unit tests are failing on Travis and I don't understand why. They pass on my desktop machine. The exception being thrown is from a check in `ParallelTransformIdentity` on a string `coordinate_system`, which if it was read from the `localmesh` should be `field_aligned`. The test should be skipped if the variable is not found in the `GridDataSource* Mesh::source`. I guess this has changed because of adding `FakeGridDataSource` (previously the check was skipped by the condition `source != nullptr` being false. `FakeGridDataSource::get(...)` always returns `false`, so `mesh.get(coordinates_type, "coordinates_type")` should return `1` (note `Mesh::get` returns an `int` with 0 for success and 1 for failure) and `!mesh.get(coordinates_type, "coordinates_type")` should evaluate to `false` skipping the check. I'm out of ideas on how to track this down... Can anyone reproduce the error or suggest a strategy for debugging? Is there a recipe to re-produce the Travis setup, e.g. in a Docker container?

**Comment by johnomotani** [comment]
These tests pass on Marconi too, with both gnu and intel compilers. :confused:

**Comment by dschwoerer** [comment]
In the past I used a docker image, starting from that old ubuntu ...

I guess your desktop is fedora, and Marconi is also CentOS - thus switching to ubuntu might do the trick ...

**Comment by d7919** [comment]
There can sometimes be (unintended) order related dependencies can you try something like

`./serial_tests --gtest_repeat=0 --gtest_shuffle --gtest_random_seed=0`

to keep on trying the tests in a different order.

**Comment by dschwoerer** [comment]
In the past I used a docker image, starting from that old ubuntu ...

I guess your desktop is fedora, and Marconi is also CentOS - thus switching to ubuntu might do the trick ...

I tried to reproduce with this:
```
podman pull ubuntu:16.04
podman run --cap-add=SYS_PTRACE --security-opt seccomp=unconfined -it --rm ubuntu:16.04 bash

apt-get	update && \
apt-get install -y mpich libmpich-dev libfftw3-dev libnetcdf-dev libnetcdf-cxx-legacy-dev git g++ make \
         python3 python3-pip python3-numpy python3-netcdf4 python3-scipy && \
pip3 install Cython && \
git clone https://github.com/boutproject/BOUT-dev && \
cd BOUT-dev && \
git checkout 41c23b0 &&	\
git submodule update --init --recursive && \
./configure && make -j 4 build-check &&	make check
```

but that fails with
```
  Compiling  nc_format.cxx
In file included from nc_format.cxx:24:0:
nc_format.hxx:115:8: error: 'bool NcFormat::read_perp(BoutReal*, const string&, int, int)' cannot be overloaded
   bool read_perp(BoutReal *var, const std::string &name, int lx = 1, int lz = 0) override;
        ^
nc_format.hxx:101:8: error: with 'bool NcFormat::read_perp(BoutReal*, const string&, int, int)'
   bool read_perp(BoutReal *var, const std::string &name, int lx = 1, int lz = 0) override;
        ^
nc_format.hxx:121:8: error: 'bool NcFormat::write_perp(BoutReal*, const string&, int, int)' cannot be overloaded
   bool write_perp(BoutReal *var, const std::string &name, int lx = 0, int lz = 0) override;
        ^
nc_format.hxx:107:8: error: with 'bool NcFormat::write_perp(BoutReal*, const string&, int, int)'
   bool write_perp(BoutReal *var, const std::string &name, int lx = 0, int lz = 0) override;
        ^
nc_format.cxx: In member function 'virtual bool NcFormat::addVarFieldPerp(const string&, bool)':
nc_format.cxx:493:81: error: no matching function for call to 'NcFile::add_var(const char*, NcType&, int, <brace-enclosed initializer list>)'
       var = dataFile->add_var(name.c_str(), nc_float_type, 3, {tDim, xDim, zDim});
                                                                                 ^
In file included from nc_format.hxx:51:0,
                 from nc_format.cxx:24:
/usr/include/netcdfcpp.h:73:20: note: candidate: virtual NcVar* NcFile::add_var(NcToken, NcType, const NcDim*, const NcDim*, const NcDim*, const NcDim*, const NcDim*)
     virtual NcVar* add_var( NcToken varname, NcType type,       // scalar
                    ^
/usr/include/netcdfcpp.h:73:20: note:   no known conversion for argument 4 from '<brace-enclosed initializer list>' to 'const NcDim*'
/usr/include/netcdfcpp.h:79:20: note: candidate: virtual NcVar* NcFile::add_var(NcToken, NcType, int, const NcDim**)
     virtual NcVar* add_var( NcToken varname, NcType type,       // n-dim
                    ^
/usr/include/netcdfcpp.h:79:20: note:   no known conversion for argument 4 from '<brace-enclosed initializer list>' to 'const NcDim**'
nc_format.cxx:495:75: error: no matching function for call to 'NcFile::add_var(const char*, NcType&, int, <brace-enclosed initializer list>)'
       var = dataFile->add_var(name.c_str(), nc_float_type, 2, {xDim, zDim});
                                                                           ^
In file included from nc_format.hxx:51:0,
                 from nc_format.cxx:24:
/usr/include/netcdfcpp.h:73:20: note: candidate: virtual NcVar* NcFile::add_var(NcToken, NcType, const NcDim*, const NcDim*, const NcDim*, const NcDim*, const NcDim*)
     virtual NcVar* add_var( NcToken varname, NcType type,       // scalar
                    ^
/usr/include/netcdfcpp.h:73:20: note:   no known conversion for argument 4 from '<brace-enclosed initializer list>' to 'const NcDim*'
/usr/include/netcdfcpp.h:79:20: note: candidate: virtual NcVar* NcFile::add_var(NcToken, NcType, int, const NcDim**)
     virtual NcVar* add_var( NcToken varname, NcType type,       // n-dim
                    ^
/usr/include/netcdfcpp.h:79:20: note:   no known conversion for argument 4 from '<brace-enclosed initializer list>' to 'const NcDim**'
nc_format.cxx: In member function 'virtual bool NcFormat::read_perp(BoutReal*, const string&, int, int)':
nc_format.cxx:618:36: error: no matching function for call to 'NcFile::get_var(const string&)'
   if(!(var = dataFile->get_var(name))) {
                                    ^
In file included from nc_format.hxx:51:0,
                 from nc_format.cxx:24:
/usr/include/netcdfcpp.h:60:12: note: candidate: NcVar* NcFile::get_var(NcToken) const
     NcVar* get_var( NcToken ) const;       // variable by name
            ^
/usr/include/netcdfcpp.h:60:12: note:   no known conversion for argument 1 from 'const string {aka const std::__cxx11::basic_string<char>}' to 'NcToken {aka const char*}'
/usr/include/netcdfcpp.h:64:12: note: candidate: NcVar* NcFile::get_var(int) const
     NcVar* get_var( int ) const;           // n-th variable
            ^
/usr/include/netcdfcpp.h:64:12: note:   no known conversion for argument 1 from 'const string {aka const std::__cxx11::basic_string<char>}' to 'int'
nc_format.cxx: In member function 'virtual bool NcFormat::write_perp(BoutReal*, const string&, int, int)':
nc_format.cxx:748:17: error: no matching function for call to 'NcFormat::checkName(const string&)'
   checkName(name);
                 ^
In file included from nc_format.cxx:24:0:
nc_format.hxx:156:8: note: candidate: void NcFormat::checkName(const char*)
   void checkName(const char* name); ///< Check if a name contains invalid characters
        ^
nc_format.hxx:156:8: note:   no known conversion for argument 1 from 'const string {aka const std::__cxx11::basic_string<char>}' to 'const char*'
nc_format.cxx:759:36: error: no matching function for call to 'NcFile::get_var(const string&)'
   if(!(var = dataFile->get_var(name))) {
                                    ^
In file included from nc_format.hxx:51:0,
                 from nc_format.cxx:24:
/usr/include/netcdfcpp.h:60:12: note: candidate: NcVar* NcFile::get_var(NcToken) const
     NcVar* get_var( NcToken ) const;       // variable by name
            ^
/usr/include/netcdfcpp.h:60:12: note:   no known conversion for argument 1 from 'const string {aka const std::__cxx11::basic_string<char>}' to 'NcToken {aka const char*}'
/usr/include/netcdfcpp.h:64:12: note: candidate: NcVar* NcFile::get_var(int) const
     NcVar* get_var( int ) const;           // n-th variable
            ^
/usr/include/netcdfcpp.h:64:12: note:   no known conversion for argument 1 from 'const string {aka const std::__cxx11::basic_string<char>}' to 'int'
nc_format.cxx:760:109: warning: format '%s' expects argument of type 'char*', but argument 3 has type 'const string* {aka const std::__cxx11::basic_string<char>*}' [-Wformat=]
    output_error.write("ERROR: NetCDF BoutReal variable '%s' has not been added to file '%s'\n", name, fname);
                                                                                                            ^
nc_format.cxx: At global scope:
nc_format.cxx:883:85: error: no 'bool NcFormat::read_rec_perp(BoutReal*, const string&, int, int)' member function declared in class 'NcFormat'
 bool NcFormat::read_rec_perp(BoutReal *data, const std::string& name, int lx, int lz) {
                                                                                     ^
nc_format.cxx:1040:86: error: no 'bool NcFormat::write_rec_perp(BoutReal*, const string&, int, int)' member function declared in class 'NcFormat'
 bool NcFormat::write_rec_perp(BoutReal *data, const std::string& name, int lx, int lz) {
                                                                                      ^
nc_format.cxx: In member function 'virtual void NcFormat::setAttribute(const string&, const string&, BoutReal)':
nc_format.cxx:1194:65: warning: format '%d' expects argument of type 'int', but argument 5 has type 'BoutReal {aka double}' [-Wformat=]
           attrname.c_str(), varname.c_str(), value, existing_att);
                                                                 ^
../../../../make.config:307: recipe for target 'nc_format.o' failed
make[4]: *** [nc_format.o] Error 1
../../../make.config:128: recipe for target 'netcdf' failed
make[3]: *** [netcdf] Error 2
../../make.config:128: recipe for target 'impls' failed
make[2]: *** [impls] Error 2
../make.config:128: recipe for target 'fileio' failed
make[1]: *** [fileio] Error 2
make.config:128: recipe for target 'src' failed
make: *** [src] Error 2
```

**Comment by johnomotani** [comment]
Thanks for the suggestions.
Shuffling the tests didn't give me any failures.
My desktop is Mint (Ubuntu-based). I guess older compiler versions or something like that might be to blame though, my system is newer than 16.04...

**Comment by dschwoerer** [comment]
After merging nc-format - I could only observe this test failing once, after running the unit tests dozens of times with `--gtest-shuffle`, only one random failure:
```
[ RUN      ] TimerTest.GetTimeLabelInScope
sys/test_timer.cxx:44: Failure
The difference between Timer::getTime("GetTimeLabelInScope test") and elapsed.count() is 0.0006093959999999999, which exceeds bout::testing::TimerTolerance, where
Timer::getTime("GetTimeLabelInScope test") evaluates to 0.00204616,
elapsed.count() evaluates to 0.0014367640000000001, and
bout::testing::TimerTolerance evaluates to 0.00050000000000000001.
[  FAILED  ] TimerTest.GetTimeLabelInScope (3 ms)
```
and the making the pid file in the root directory are the only errors I got ...

This is (roughly) what I used:
```
podman pull ubuntu:16.04
podman run -it --rm ubuntu:16.04 bash

apt-get update && \
apt-get install -y mpich libmpich-dev libfftw3-dev libnetcdf-dev libnetcdf-cxx-legacy-dev git g++ make \
         python3 python3-pip python3-numpy python3-netcdf4 python3-scipy gdb && \
pip3 install Cython && \
git clone https://github.com/boutproject/BOUT-dev && \
cd BOUT-dev && \
git checkout 41c23b0 && \
git merge origin/nc-format || git commit -m 'fu' && \
git submodule update --init --recursive && \
export LANG=C.UTF-8 && \
echo " $(hostname)" >> /etc/hosts && \
./configure && make -j 4 build-check && cd tests/unit && make check
```

**Comment by ZedThree** [comment]
I'm struggling to reproduce this even on a ubuntu 16 machine, although on a slightly older kernel. It would be quite surprising if that were to blame though.

I've not found an actual image for the exact system Travis is using, but I'll see if a vanilla ubuntu 16 VM gets me any further

File: src/mesh/coordinates.cxx [hunk 1]
```diff
@@ -275,18 +275,9 @@ Coordinates::Coordinates(Mesh* mesh, Options* options)
...
-      || (min(g33, RGN_NOBNDRY) <= 0.0)) {
-    throw BoutException("\tERROR: Diagonal metrics are negative!\n");
-  }
-  if ((!finite(g12, RGN_NOBNDRY)) || (!finite(g13, RGN_NOBNDRY))
-      || (!finite(g23, RGN_NOBNDRY))) {
-    throw BoutException("\tERROR: Off-diagonal metrics are not finite!\n");
-  }
+  checkFinite(g11, "g11", "RGN_NOCORNERS");
+  checkFinite(g22, "g22", "RGN_NOCORNERS");
+  checkFinite(g33, "g33", "RGN_NOCORNERS");
```
Comment (ZedThree) [inline src/mesh/coordinates.cxx]: Is it on purpose that only the diagonal metrics are checked here, and only checked for finite-ness?

**Comment by ZedThree** [comment]
I cannot reproduce this bug even with the identical kernel, so it must be something unique about the Travis system. It also can't be a compiler bug, because it happens to clang as well. I'll try a couple of other things, and then I think I'll need to ask Travis to enable debug mode so I can get in and run gdb on the actual VM.

File: src/mesh/coordinates.cxx (hunk 1 above)
Comment (johnomotani) [reply src/mesh/coordinates.cxx]: Just a brain-freeze! thanks for spotting :+1:

File: include/field.hxx [hunk 2]
```diff
@@ -269,4 +271,42 @@ inline T filledFrom(const T& f, BoutReal fill_value) {
...
 T operator+(const T& f) {return f;}
 
+/// Check if all values of a field \p var are finite.  Loops over all points including the
+/// boundaries by default (can be changed using the \p rgn argument)
+/// If any element is not finite, throws an exception that includes the position of the
+/// first found.
+template<typename T>
+inline void checkFinite(const T& f, const std::string& name="field", const std::string& rgn="RGN_ALL") {
+  AUTO_TRACE();
+
```
Comment (ZedThree) [inline include/field.hxx]: With #1730 this could include `static_assert(bout::utils::is_Field<T>::value, "checkFinite is only valid for Fields")`

**Comment by ZedThree** [comment]
`checkFinite` and `checkPositive` are nice. I'm a little concerned that `checkFinite` is almost identical to `checkData` though. They clearly have slightly different uses -- `checkData` may be a no-op, but `checkFinite` always checks. I'm not sure what, if anything, we should do about that though.

`checkFinite/Positive` could be renamed perhaps, though I don't know what to. Maybe best to just add something to the docstring for now saying that they always check.

Also, they should probably go into a namespace -- `bout::`, at least. We don't really have a good feel for sub-namespaces at the minute, most things are going into `bout::utils::` but we can't have everything in there, so that might be something to sort out for 5.0

**Pull request merged**